python-package/xgboost/sklearn.py

```diff
@@ -898,13 +935,30 @@ def get_params(self, deep: bool = True) -> Dict[str, Any]:
         """Get parameters."""
         # Based on: https://stackoverflow.com/questions/59248211
         # The basic flow in `get_params` is:
-        # 0. Return parameters in subclass first, by using inspect.
-        # 1. Return parameters in `XGBModel` (the base class).
+        # 0. Return parameters in subclass (self.__class__) first, by using inspect.
+        # 1. Return parameters in all parent classes (especially `XGBModel`).
         # 2. Return whatever in `**kwargs`.
         # 3. Merge them.
+        #
+        # This needs to accommodate being called recursively in the following
+        # inheritance graphs (and similar for classification and ranking):
+        #
+        #   XGBRFRegressor -> XGBRegressor -> XGBModel -> BaseEstimator
+        #                     XGBRegressor -> XGBModel -> BaseEstimator
+        #                                     XGBModel -> BaseEstimator
+        #
         params = super().get_params(deep)
         cp = copy.copy(self)
-        cp.__class__ = cp.__class__.__bases__[0]
+        # if the immediate parent is a mixin, skip it (mixins don't define get_params())
```

Comment:
Do you think it's more general to check for the `get_params` attribute instead of checking hardcoded mixins? The current check seems to defeat the purpose of having a polynormal structure (inheritance).

Reply:
Yes that's a good idea! Just pushed https://github.com/dmlc/xgboost/pull/11021/commits/18c602f96e903c46b1145d62e9ffd2d9ae5a4c78 implementing that.